tests/FakeItEasy.Tests/ConcurrentCallTests.cs

```diff
@@ -0,0 +1,59 @@
+namespace FakeItEasy.Tests
+{
+    using System.Threading;
+    using System.Threading.Tasks;
+    using FluentAssertions;
+    using Xunit;
+    using Xunit.Abstractions;
+
+    public class ConcurrentCallTests
+    {
+        private readonly ITestOutputHelper output;
+
+        public ConcurrentCallTests(ITestOutputHelper output)
+        {
+            this.output = output;
+        }
+
+        public interface ITestInterface
+        {
+            int MyMethod();
+        }
+
+        [Fact]
+        public void Concurrent_calls_are_correctly_recorded()
+        {
+            for (int i = 0; i < 100; i++)
+            {
+                try
+                {
+                    ITestInterface fake = A.Fake<ITestInterface>();
+
+                    int count = 0;
+                    A.CallTo(() => fake.MyMethod()).ReturnsLazily(p => Interlocked.Increment(ref count));
+
+                    Task<int>[] tasks =
+                    {
+                        Task.Run(() => fake.MyMethod()),
+                        Task.Run(() => fake.MyMethod()),
+                    };
+
+                    int[] result = Task.WhenAll(tasks).Result;
+
+                    result.Should().BeEquivalentTo(1, 2);
+
+                    // Assertion fails NEVER with FakeItEasy 2.3.1
+                    // Assertion fails sometimes with FakeItEasy 4.7.1
+                    // Assertion fails sometimes with FakeItEasy 4.9.0
+                    // Message: "Expected to find it twice exactly but found it once among the calls: ..."
+                    A.CallTo(() => fake.MyMethod()).MustHaveHappened(Repeated.Exactly.Twice);
+                }
+                catch
+                {
+                    this.output.WriteLine($"Failed at iteration {i}");

```

Comment:
I included this for convenience, but I can remove it.

Reply:
No, I think if it fails, we'll enjoy having the count!